src/main/java/com/redhat/pantheon/servlet/AbstractJsonSingleQueryServlet.java

```diff
@@ -0,0 +1,78 @@
+package com.redhat.pantheon.servlet;
+
+import com.redhat.pantheon.jcr.JcrQueryHelper;
+import com.redhat.pantheon.model.QuerySingleResult;
+import org.apache.http.entity.ContentType;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.api.SlingHttpServletResponse;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.servlets.SlingSafeMethodsServlet;
+import org.jetbrains.annotations.NotNull;
+
+import javax.annotation.Nonnull;
+import javax.jcr.RepositoryException;
+import javax.servlet.ServletException;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.google.common.collect.Maps.newHashMap;
+import static com.redhat.pantheon.servlet.ServletUtils.writeAsJson;
+import static javax.servlet.http.HttpServletResponse.SC_NOT_FOUND;
+
+/**
+ * Servlet that helps create simple Json result from custom queries.
+ * The servlet returns the first object from the query response irrespective
+ * of the number of elements returned in the response.
+ * This servlet should be extended in the case where you need only the 1st/single element
+ * from the result in json format. See methods to be overriden for details on how to
+ * customize sub classes.
+ *
+ * @author Ankit Gadgil
+ */
+public abstract class AbstractJsonSingleQueryServlet extends SlingSafeMethodsServlet {
+
+    /**
+     * Returns the query to execute. The query may be modified depending on the provided
+     * parameters in the request.
+     * @param request The sling servlet request
+     * @return A string with the query to execute. The results of this query will be used
+     * to render the list of results.
+     */
+    protected abstract String getQuery(SlingHttpServletRequest request);
+
+    /**
+     * Provides a way to modify the returned objects based on the found resources.
+     * The default implementation just returns the corresponding value map.
+     * @param resource The Resource obtained as a result of the query.
+     * @return A map with the actual value to be returned to the servlet's caller.
+     */
+    protected Map<String, Object> resourceToMap(@NotNull Resource resource) {
+        return newHashMap(resource.getValueMap());
+    }
+
+    @Override
+    protected final void doGet(@Nonnull SlingHttpServletRequest request, @Nonnull SlingHttpServletResponse response)
+            throws ServletException, IOException {
+
+        JcrQueryHelper queryHelper = new JcrQueryHelper(request.getResourceResolver());
+        try {
+            Stream<Resource> resultStream = queryHelper.query(getQuery(request));
+            List<Map<String, Object>> resultList = resultStream.map(this::resourceToMap)
```

Comment:
you only need to transform the first resource, not all of them. I would move this call to after you have extracted the first result.

Reply:
I think, the idea is that yes it should return a single result (we have another abstract servlet for multiple results).

As to throwing an exception if there is more than one result... I’m not sure. It depends how much responsibilities we want to give the abstract class. We could leave said responsibilities to the sub-classes, or hard-code them here.

Reply:
Perhaps we have a `protected Resource handleMultipleQueryResults(Stream<Resource> results)` method in the abstract class which by default just throws an exception in the face of multiple results. Then subclasses can override if/when it makes sense. That would be my suggestion.